clustermesh-apiserver/syncstate/syncstate.go

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package syncstate
+
+import (
+	"context"
+
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/metrics/metric"
+	"github.com/cilium/cilium/pkg/spanstat"
+)
+
+var Cell = cell.Module(
+	"sync",
+	"ClusterMesh Sync",
+
+	cell.Metric(MetricsProvider()),
+	cell.Provide(new),
+)
+
+func new(metrics Metrics) SyncState {
+	ss := SyncState{StoppableWaitGroup: lock.NewStoppableWaitGroup()}
+	syncTime := spanstat.Start()
+
+	go func() {
+		<-ss.WaitChannel()
+		metrics.InitialSyncDuration.Set(syncTime.Seconds())
+	}()
+	return ss
+}
+
+// SyncState is a wrapper around lock.StoppableWaitGroup used to keep track of the synchronization
+// of various resources to the kvstore.
+type SyncState struct {
+	*lock.StoppableWaitGroup
+}
+
+// Complete returns true if all resources have been synchronized to the kvstore.
+func (ss SyncState) Complete() bool {
+	select {
+	case <-ss.WaitChannel():
+		return true
+	default:
+		return false
+	}
+}
+
+// WaitForResource adds a resource to the SyncState and returns a callback function that should be
+// called when the resource has been synchronized.
+func (ss SyncState) WaitForResource() func(context.Context) {
+	ss.Add()
+	return func(_ context.Context) {
+		ss.Done()
+	}
+}
+
+// Metrics contains metrics that should only be exported by the
+// clustermesh-apiserver or kvstoremesh.
+type Metrics struct {
+	// InitialSyncDuration tracks the duration in seconds of the kvstore initial sync.
+	InitialSyncDuration metric.Gauge
+}
+
+func MetricsProvider() func() Metrics {
+	return func() Metrics {
```

Comment:
This extra function looks unnecessary. You could simply return the Metrics object.